File: Creational/Factory/DIY.java
package Creational.Factory;


public class DIY {
    public static void main(String[] args) {
        new ProductControler().listProducts();
    }
}

abstract class ControllerMods{

    public void add( String fileName){
        
        var file = createFiles();
        file.render("GameofLife.exe");

    }
    //factory
    protected abstract Files createFiles();

}

//Product controller
class ProductControler extends HairController{
    public void listProducts(){
        add("mods.exe");
    }
    
    
}

//libraries
class diffHairMod implements Files{
    public void render(String fileName) {
        System.out.println("256 hairstyles added");
    }

}

 class HairController extends ControllerMods{
    @Override
    protected Files createFiles() {
        return new diffHairMod();
    }
}

//another library 

class addWingsMod implements Files{

    public void render(String fileName) {
        System.out.println("Wings added");
    }

}

class WingsController extends ControllerMods{

    @Override
    protected Files createFiles() {
      return new addWingsMod();
    }

}

interface Files{
   void render(String fileName);
}

File: TS/Creational/Factory/DIY.ts
abstract class ControllerMods {
  public add(fileName: string): void {
    const file = this.createFiles();
    file.render("GameofLife.exe");
  }

  // Factory method
  protected abstract createFiles(): Files;
}

class HairController extends ControllerMods {
  protected createFiles(): Files {
    return new DiffHairMod();
  }
}
// Product controller
class ProductController extends HairController {
  public listProducts(): void {
    this.add("mods.exe");
  }
}

// Libraries
class DiffHairMod implements Files {
  public render(fileName: string): void {
    console.log("256 hairstyles added");
  }
}
class WingsController extends ControllerMods {
  protected createFiles(): Files {
    return new AddWingsMod();
  }
}

// Another library
class AddWingsMod implements Files {
  public render(fileName: string): void {
    console.log("Wings added");
  }
}

interface Files {
  render(fileName: string): void;
}

// Main function
function DIYFactory(): void {
  new ProductController().listProducts();
}

DIYFactory();
